### DISCUSSION TITLE
New version of MOOSE computationally slower than older version

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#17671] [Q&A General] [answered]
Dear all,
I was facing performance issues with current version of MOOSE as it seems to be slower than the older versions dated about an year ago. My plasticity model with ~5000 Dof's took nearly an hour of computational time, while the legacy code ran in roughly around 35mins. The related discussion can be traced back to [https://github.com/idaholab/moose/discussions/16372]. Unfortunately the compilers here are upgraded and hence i am unable to build the legacy sources now, which might have given a chance for me to check and re-confirm their time. Probably somebody there in the team could test the attached files (with older versions ranging around May 2020) and check if these can be run under an hour. Any insights on the possible reasons for the slower performance will be appreciated, as this may lead to a huge savings in time while conduting the iterative simulation that i am currently working upon. 
[23042021.zip](https://github.com/idaholab/moose/files/6366233/23042021.zip)

**Comment by jiangwen84** [comment]
Couple of questions:

1. could you let us the commit hash of the old MOOSE you referred to? 
2. are you running exactly the same input file and using the same number of processors? 
3. Since you use AD, you should set 'solve_type = Newton' instead of 'PJFNK'. Could you change the solver type and rerun it?

**Comment by abarun22** [comment]
Dear Wen,
Here are my answers.
              1) Could you let us the commit hash of the old MOOSE you
referred to?
                 > git log -1
                commit 1cc6b7e797570453696906969982e92d58b9c9a9
                Merge: 314e8b9 6fd9c7f
               Author: moosetest ***@***.***>
               Date:   Mon May 18 17:31:13 2020 -0600
              Merge commit '6fd9c7f846a43385430795ac9e971716d2218d1c'

             2) Are you running exactly the same input file and using the
same number of processors?
                 Yes the same input file and the same number of processors
(#10)

              3) Since you use AD, you should set 'solve_type = Newton'
instead of 'PJFNK'. Could you change the solver type and rerun it?
                  I did try with Newton, but for some reasons it wouldn't
run at that time. Will give a try again and let you know.

The functional filename is different in the inputfile and so request you to
set the actuals while running the computation.


Kind regards,
Arun

**Comment by hugary1995** [comment]
I can reproduce @abarun22 's results.

I slightly modified his input file to simplify things:

```
[Mesh]
  type = FileMesh
  file = Specimen_QtrModel.e
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  volumetric_locking_correction = true
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    incremental = true
    add_variables = true
    use_automatic_differentiation = true
  []
[]

[BCs]
  [Load_RE]
    type = FunctionDirichletBC
    variable = disp_x
    function = '0.0226*t'
    boundary = "Right_end"
    preset = false
  []
  [Symm_LE]
    type = DirichletBC
    variable = disp_x
    value = 0
    boundary = "Left_end"
  []
  [Symm_BY]
    type = DirichletBC
    variable = disp_y
    value = 0.0
    boundary = "Bottom"
  []
  [Symm_LZ]
    type = DirichletBC
    variable = disp_z
    value = 0.0
    boundary = "Lateral_face"
  []
  [Const_RY]
    type = DirichletBC
    variable = disp_y
    value = 0.0
    boundary = "Right_end"
  []
  [Const_RZ]
    type = DirichletBC
    variable = disp_z
    value = 0.0
    boundary = "Right_end"
  []
[]

[Functions]
  [hf]
    type = PiecewiseLinear
    data_file = PW_last_iter.csv
    scale_factor = 1
    format = columns
  []
[]

[Materials]
  [elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 2.02E+11
    poissons_ratio = 0.315
  []
  [isotropic_plasticity]
    type = ADIsotropicPlasticityStressUpdate
    yield_stress = 554146341.5
    hardening_function = hf
  []
  [radial_return_stress]
    type = ADComputeMultipleInelasticStress
    inelastic_models = 'isotropic_plasticity'
  []
[]

[Postprocessors]
  [nl_its]
    type = NumNonlinearIterations
  []
  [time]
    type = PerfGraphData
    section_name = Root
    data_type = TOTAL
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu       superlu_dist'

  dt = 0.1
  end_time = 25

  nl_rel_tol = 1e-6
  nl_max_its = 20

  automatic_scaling = true
[]

[Outputs]
  exodus = true
  csv = true
  print_linear_residuals = false
[]
```

I use the same conda env to make a more fair comparison. Most important versions are:
```
moose-mpich               3.3.2                   build_5    idaholab
moose-petsc               3.14.2                  build_4    idaholab
moose-tools               2021.01.06                 py_0    idaholab
```

New moose (c7126c855641b1f318fc64faa32cc8070fdf2231) total run time: 449.5038 seconds
Old moose (1cc6b7e797570453696906969982e92d58b9c9a9) total run time: 296.3510 seconds

This may well be due to a simple config change or default option change. But I think there is value in tracking it down.

**Reply by hugary1995** [reply]
BTW, this is using 8 processors on my Ubuntu.

**Reply by YaqiWang** [reply]
Which commit is causing this?

**Reply by hugary1995** [reply]
I don't know. I haven't run any bisect nor any profiling yet.

**Reply by lindsayad** [reply]
Gary can you see whether changing to nonsparse changes things? (I’m guessing you were running with default sparse crowd config?)

**Reply by jiangwen84** [reply]
> I don't know. I haven't run any bisect nor any profiling yet.

@hugary1995 There are definitely some changes in`ADComputeMultipleInelasticStress`, which also might cause this issue.

**Comment by hugary1995** [comment, marked as answer]
@lindsayad Yeah I was running with default config. When I try to switch to nonsparse local indexing I get the following error:
```
$> ./configure --with-ad-indexing-type=local                                                                                             ─╯
---------------------------------------------
----------- Configuring MOOSE ---------------
---------------------------------------------
configuring with derivative backing array size of 50
configuring with AD indexing type local
configuring with derivative type sparse
checking for x86_64-conda-linux-gnu-pkg-config... no
checking for pkg-config... /usr/bin/pkg-config
configure: checking system for libpng
configuring with libpng support
configure: creating ./config.status
config.status: creating conf_vars.mk
config.status: creating framework/include/base/MooseConfig.h.tmp
config.status: executing framework/include/base/MooseConfig.h commands
config.status: creating framework/include/base/MooseConfig.h - prefix MOOSE for framework/include/base/MooseConfig.h.tmp defines
---------------------------------------------
--------- Done Configuring MOOSE ------------
---------------------------------------------
```

```
$> ./configure --with-derivative-type=nonsparse                                                                                          ─╯
---------------------------------------------
----------- Configuring MOOSE ---------------
---------------------------------------------
configuring with derivative backing array size of 50
configure: error: cannot use global AD indexing with a nonsparse derivative container
```

Do you have any clue?

**Reply by hugary1995** [reply]
Nvmd, I need to do `./configure --with-derivative-type=nonsparse --with-ad-indexing-type=local `. I thought the configure script knows about the current config status.

**Reply by hugary1995** [reply]
@lindsayad  Good news for me: I don't need to run any bisects :)
Bad news for you: with AD local indexing nonsparse the run time is 287.7488, with AD global indexing sparse the run time is 440.2642 (Yes, the number is different from the one in my original post because I re-ran it to make sure.)

**Reply by hugary1995** [reply]
@abarun22 Could you try to run `./configure --with-derivative-type=nonsparse --with-ad-indexing-type=local` in your moose root directory, then recompile and see if it speeds up your calculations?

**Reply by lindsayad** [reply]
Ok I’m glad we figured out the issue straightaway. I will have to profile this to figure out the differences and whether there are any potential remedies. In the meantime the user should definitely work with local nonsparse. Gary could you create a ticket and tag me please?

**Reply by hugary1995** [reply]
Thanks! No problem, I just tagged you.

**Reply by abarun22** [reply]
Dear Gary,
The configure options for non-sparse indexing does not seem to impact the computational time. I ran your version of the input file which took 462s and 466s before and after configuring the options. It is still taking the same one hour for my version of input file. 

FYI, i did the configuration as suggested by you, which it did successfully - then moved to test directory (moose/test), recompiled the sources with new options and recompiled my application eventully. 
Kind regards,
Arun

**Reply by hugary1995** [reply]
You made me unconfident... I triple checked just now, and global sparse gave me 434 seconds and local nonsparse gave me 290 seconds. It is really making a difference on my machine. Either one of us did something wrong, or something is machine-dependent. We probably need a third person to confirm.

Just to make sure, you first ran the `./configure --with-derivative-type=nonsparse --with-ad-indexing-type=local` command, then recompiled your application, and finally ran my input file, right? Is it possible you have multiple moose on your machine and your application didn't pick up the right one?

**Reply by abarun22** [reply]
Dear Gary,
Yes exactly the same. Here is the configure log for your reference
(base) abalasub@freia013> ./configure --with-derivative-type=nonsparse --with-ad-indexing-type=local
---------------------------------------------
----------- Configuring MOOSE ---------------
---------------------------------------------
configuring with derivative backing array size of 50
configuring with AD indexing type local
configuring with derivative type nonsparse
checking for pkg-config... /usr/bin/pkg-config
configure: checking system for libpng
configuring with libpng support
configure: creating ./config.status
config.status: creating conf_vars.mk
config.status: creating framework/include/base/MooseConfig.h.tmp
config.status: executing framework/include/base/MooseConfig.h commands
config.status: creating framework/include/base/MooseConfig.h - prefix MOOSE for framework/include/base/MooseConfig.h.tmp defines
---------------------------------------------
--------- Done Configuring MOOSE ------------

After that i re-compiled my application and ran your inputfile. I am sure the application is linking with the correct version of moose uniquely. No problems here i guess. I will try and run this nonsparse configured moose in other clusters and see if it speeds up the computation.
